README.md
# node-dependency-parser
Run `yarn` to download dependencies, `nodegit` may take a while to install

## To Test
Modify `package.json` `dependencies` or `devDependencies` or add a `package.json` in a subdirectory and commit changes. 
Then run `yarn dev` to write dependency changes to `CHANGELOG.md`.


CHANGELOG.md
# Latest Commit: updated file structure 
## Dependency Changes 
#### No dependencies changed
      

src/index.ts
import { Repository } from "nodegit";
import path from "path";
import {
  getJsonFieldBounds,
  updatePatchChanges,
  parseDependencyDiff,
  writeChangelog,
} from "./helpers";
//Removed dependencies, added dependencies, updated dependencies (versions)

export interface Dependency {
  name: string;
  version: string;
  type: "removed" | "added" | "updated";
  prevVersion?: string;
}

export interface DependencyMapping {
  [name: string]: Dependency;
}

export interface DependencyDiff {
  removed: Dependency[];
  added: Dependency[];
  updated: Dependency[];
}

export interface FileDiff {
  fileName: string;
  dependencies: DependencyDiff | null;
  devDependencies: DependencyDiff | null;
}

async function parseCommitDependenciesDiff() {
  //map file path to added, deleted, and/or updateed dependencies
  const commitDependenciesDiffs: FileDiff[] = [];
  const repo = await Repository.open(path.resolve(__dirname, "../.git"));
  const currentBranchReference = await repo.getCurrentBranch();
  const latestCommit = await repo.getBranchCommit(currentBranchReference);
  const diffList = await latestCommit.getDiff();
  for (const diff of diffList) {
    const patches = await diff.patches();

    for (const patch of patches) {
      const fileName = patch.newFile().path().split("/").at(-1);
      if (fileName == "package.json") {
        const patchChanges: DependencyMapping = {};
        const patchDevChanges: DependencyMapping = {};

        //get line ranges for depencies and devDependencies
        const dependenciesBounds = await getJsonFieldBounds(
          patch.newFile().path(),
          `"dependencies`
        );
        const devDependenciesBounds = await getJsonFieldBounds(
          patch.newFile().path(),
          `"devDependencies"`
        );

        const hunks = await patch.hunks();
        for (const hunk of hunks) {
          const lines = await hunk.lines();
          let lineCount = hunk.newStart() - 1;

          for (const line of lines) {
            //line.origin() returns an ASCII charCode, 43 is a +, 45 is a -
            if (line.origin() != 45) {
              lineCount++;
            }

            //lines are within dependencies or devDepencies range
            if (
              dependenciesBounds.start <= lineCount &&
              dependenciesBounds.end >= lineCount &&
              !line.content().trim().includes(`"dependencies`) &&
              !line.content().trim().includes("}")
            ) {
              updatePatchChanges(patchChanges, line);
            } else if (
              devDependenciesBounds.start <= lineCount &&
              devDependenciesBounds.end >= lineCount &&
              !line.content().trim().includes(`"devDependencies`) &&
              !line.content().trim().includes("}")
            ) {
              updatePatchChanges(patchDevChanges, line);
            }
          }
        }

        const dependenciesDiff = parseDependencyDiff(patchChanges);
        const devDependenciesDiff = parseDependencyDiff(patchDevChanges);

        if (dependenciesDiff || devDependenciesDiff) {
          commitDependenciesDiffs.push({
            fileName: patch.newFile().path().trim(),
            dependencies: dependenciesDiff,
            devDependencies: devDependenciesDiff,
          });
        }
      }
    }
  }
  writeChangelog(latestCommit, commitDependenciesDiffs);
}

parseCommitDependenciesDiff();


src/helpers/write-changelog.helper.ts
import { createWriteStream } from "fs";
import { Commit } from "nodegit";
import { Dependency, DependencyDiff, FileDiff } from "..";

function getSectionMarkdown(
  section: Dependency[],
  header: string,
  isUpdate?: boolean
) {
  if (section.length < 1) {
    return ``;
  }
  return `>##### ${header}\n${section
    .map(
      (dep) =>
        `>* \`${dep.name}: ${dep.prevVersion}\`${
          isUpdate ? `\`=> ${dep.version}\`` : ``
        }`
    )
    .join(`\n`)}
    `;
}

function getDependenciesMarkdown(
  dependencies: DependencyDiff | null,
  title: string
) {
  if (!dependencies) {
    return ``;
  }
  const header = `#### ${title} \n`;
  const added = getSectionMarkdown(dependencies.added, "Added");
  const removed = getSectionMarkdown(dependencies.removed, "Removed");
  const updated = getSectionMarkdown(dependencies.updated, "Updated", true);
  return `${header}${added}${removed}${updated}`.trim();
}

function getFileMarkdown(file: FileDiff) {
  const header = `### \`${file.fileName}\` \n`;
  const dependencies = getDependenciesMarkdown(
    file.dependencies,
    "Dependencies"
  );
  const devDependencies = getDependenciesMarkdown(
    file.devDependencies,
    "Dev Dependencies"
  );
  return `${header}${dependencies}\n${devDependencies}\n`;
}

export async function writeChangelog(
  latestCommit: Commit,
  commitDependenciesDiffs: FileDiff[]
) {
  var stream = createWriteStream("CHANGELOG.md", { flags: "w" });
  stream.write(
    `# Latest Commit: ${latestCommit
      .message()
      .trim()} \n## Dependency Changes \n${
      commitDependenciesDiffs.length > 0
        ? commitDependenciesDiffs.map((file) => getFileMarkdown(file)).join(``)
        : "#### No dependencies changed"
    }
      `
  );

  stream.end();
}


src/helpers/update-patch-changes.helper.ts
import { DiffLine } from "nodegit";
import { DependencyMapping } from "..";

export function updatePatchChanges(
  patchChanges: DependencyMapping,
  line: DiffLine
): void {
  const parsedLine = line.content().trim().split(" ");
  const [name, version] = parsedLine.map((val) =>
    val.at(-1) == "," || val.at(-1) == ":" ? val.slice(0, -1) : val
  );
  //line was removed
  if (line.origin() == 45) {
    if (name in patchChanges && patchChanges[name].type == "added") {
      if (patchChanges[name].version == version) {
        delete patchChanges[name];
      } else {
        patchChanges[name].type = "updated";
        patchChanges[name].prevVersion = version;
      }
    } else if (!(name in patchChanges)) {
      patchChanges[name] = {
        name,
        version,
        type: "removed",
      };
    }
  }
  //line was added
  else if (line.origin() == 43) {
    if (name in patchChanges && patchChanges[name].type == "removed") {
      if (patchChanges[name].version == version) {
        delete patchChanges[name];
      } else {
        patchChanges[name].type = "updated";
        patchChanges[name].prevVersion = patchChanges[name].version;
        patchChanges[name].version = version;
      }
    } else if (!(name in patchChanges)) {
      patchChanges[name] = {
        name,
        version,
        type: "added",
      };
    }
  }
}


src/helpers.ts
import { createReadStream, createWriteStream } from "fs";
import { Commit, DiffLine } from "nodegit";
import { createInterface } from "readline";
import {
  CommitDiffMapping,
  Dependency,
  DependencyDiff,
  DependencyMapping,
  FileDiff,
} from ".";

export async function getJsonFieldBounds(
  filePath: string,
  fieldName: string
): Promise<{ start: number; end: number }> {
  const fileStream = createReadStream(filePath);

  const rl = createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });
  let lineNum = 0;
  let start = 0;
  let insideField = false;
  for await (const line of rl) {
    // Each line in input.txt will be successively available here as `line`.
    lineNum += 1;
    if (line.includes(fieldName)) {
      start = lineNum;
      insideField = true;
    }
    if (insideField && line.includes("}")) {
      return { start, end: lineNum };
    }
  }
  throw `fieldName ${fieldName} not found in ${filePath}`;
}

export function updatePatchChanges(
  patchChanges: DependencyMapping,
  line: DiffLine
): void {
  const parsedLine = line.content().trim().split(" ");
  const [name, version] = parsedLine.map((val) =>
    val.at(-1) == "," || val.at(-1) == ":" ? val.slice(0, -1) : val
  );
  //line was removed
  if (line.origin() == 45) {
    if (name in patchChanges && patchChanges[name].type == "added") {
      if (patchChanges[name].version == version) {
        delete patchChanges[name];
      } else {
        patchChanges[name].type = "updated";
        patchChanges[name].prevVersion = version;
      }
    } else if (!(name in patchChanges)) {
      patchChanges[name] = {
        name,
        version,
        type: "removed",
      };
    }
  }
  //line was added
  else if (line.origin() == 43) {
    if (name in patchChanges && patchChanges[name].type == "removed") {
      if (patchChanges[name].version == version) {
        delete patchChanges[name];
      } else {
        patchChanges[name].type = "updated";
        patchChanges[name].prevVersion = patchChanges[name].version;
        patchChanges[name].version = version;
      }
    } else if (!(name in patchChanges)) {
      patchChanges[name] = {
        name,
        version,
        type: "added",
      };
    }
  }
}

export function parseDependencyDiff(
  patchChanges: DependencyMapping
): DependencyDiff | null {
  const dependencyDiff: DependencyDiff = {
    removed: [],
    added: [],
    updated: [],
  };
  if (
    Object.keys(patchChanges).length === 0 &&
    patchChanges.constructor === Object
  ) {
    return null;
  }
  for (const [name, dependency] of Object.entries(patchChanges)) {
    dependencyDiff[dependency.type].push(dependency);
  }
  return dependencyDiff;
}

function getSectionMarkdown(
  section: Dependency[],
  header: string,
  isUpdate?: boolean
) {
  if (section.length < 1) {
    return ``;
  }
  return `>##### ${header}\n${section
    .map(
      (dep) =>
        `>* \`${dep.name}: ${dep.version}\`${
          isUpdate ? `\`=> ${dep.prevVersion}\`` : ``
        }`
    )
    .join(`\n`)}
  `;
}

function getDependenciesMarkdown(
  dependencies: DependencyDiff | null,
  title: string
) {
  if (!dependencies) {
    return ``;
  }
  const header = `#### ${title} \n`;
  const added = getSectionMarkdown(dependencies.added, "Added");
  const removed = getSectionMarkdown(dependencies.removed, "Removed");
  const updated = getSectionMarkdown(dependencies.updated, "Updated", true);
  return `${header}${added}${removed}${updated}`.trim();
}

function getFileMarkdown(file: FileDiff) {
  const header = `### \`${file.fileName}\` \n`;
  const dependencies = getDependenciesMarkdown(
    file.dependencies,
    "Dependencies"
  );
  const devDependencies = getDependenciesMarkdown(
    file.devDependencies,
    "Dev Dependencies"
  );
  return `${header}${dependencies}\n${devDependencies}\n`;
}

export async function writeChangelog(
  latestCommit: Commit,
  commitDependenciesDiffs: FileDiff[]
) {
  var stream = createWriteStream("CHANGELOG.md", { flags: "w" });
  stream.write(
    `# Latest Commit: ${latestCommit
      .message()
      .trim()} \n## Dependency Changes \n${
      commitDependenciesDiffs.length > 0
        ? commitDependenciesDiffs.map((file) => getFileMarkdown(file)).join(``)
        : "#### No dependencies changed"
    }
    `
  );

  stream.end();
}
